File: vscode-extension/src/sections/workflows-section.ts
import * as vscode from 'vscode';
import { StatusItem } from '../status-item';
import type { SectionContext, StatusSection } from './types';
import type { SectionId, SectionHealth } from '../generated';
import { discoverApiUrl } from '../api-client';
import type { WorkflowFormatDto } from '../generated/WorkflowFormatDto';

/**
 * Workflows section — the export formats a ticket + issue type can be rendered
 * into (Claude `.js`, AGNT `.json`). Info-only (`Gray`), gated on `connections`:
 * preview/export run against the hosted API (`/api/v1/workflow-formats`), so the
 * section only appears once connections are ready. Rows link out to the hosted
 * Operator UI's Workflows page, where preview/export run — the extension does
 * not reimplement that surface.
 */
interface WorkflowsState {
  apiAvailable: boolean;
  formats: WorkflowFormatDto[];
}

export class WorkflowsSection implements StatusSection {
  readonly sectionId: SectionId = 'workflows';
  readonly prerequisites: SectionId[] = ['connections'];

  private state: WorkflowsState = { apiAvailable: false, formats: [] };

  health(): SectionHealth {
    return 'Gray';
  }

  async check(ctx: SectionContext): Promise<void> {
    try {
      const apiUrl = await discoverApiUrl(ctx.ticketsDir);
      const response = await fetch(`${apiUrl}/api/v1/workflow-formats`);
      if (response.ok) {
        const formats = await response.json() as WorkflowFormatDto[];
        this.state = { apiAvailable: true, formats };
        return;
      }
    } catch {
      // API not available — fall through to the unavailable state.
    }
    this.state = { apiAvailable: false, formats: [] };
  }

  getTopLevelItem(_ctx: SectionContext): StatusItem {
    const count = this.state.formats.length;
    return new StatusItem({
      label: 'Workflows',
      description: this.state.apiAvailable ? `${count} export formats` : 'API required',
      icon: 'type-hierarchy',
      collapsibleState: count > 0
        ? vscode.TreeItemCollapsibleState.Collapsed
        : vscode.TreeItemCollapsibleState.None,
      sectionId: this.sectionId,
      health: this.health(),
    });
  }

  getChildren(_ctx: SectionContext, _element?: StatusItem): StatusItem[] {
    return this.state.formats.map((fmt) => new StatusItem({
      label: fmt.label,
      description: `${fmt.status} · .${fmt.extension}`,
      icon: 'tools',
      collapsibleState: vscode.TreeItemCollapsibleState.None,
      sectionId: this.sectionId,
      health: this.health(),
      // Link out to the hosted UI's Workflows page (preview/export live there).
      command: { command: 'operator.openWorkflows', title: 'Open Workflows in Operator UI' },
    }));
  }
}

File: src/ui/sections/workflows_section.rs
//! The **Workflows** status section: the export formats a ticket + issuetype can
//! be rendered into (Claude dynamic workflow `.js`, AGNT graph `.json`).
//!
//! Info-only — formats carry no credentials, so the section is `Gray`. It gates
//! on `Connections`: preview/export run against the hosted API, so the section
//! only appears once connections are ready. Each row names a format, its support
//! status + file extension, and
//! links to its docs. The primary action opens the web UI's Workflows page where
//! per-issuetype preview / per-ticket export run against the existing endpoints.
//! The format list is the same source of truth the REST `workflow-formats`
//! endpoint and the `Workflows` catalog vertical derive from: [`WorkflowFormat`].

use crate::integrations::{catalog, Vertical};
use crate::ui::status_panel::{
    ActionMeta, ActionSet, SectionHealth, SectionId, StatusAction, StatusIcon, StatusSection,
    StatusSnapshot, TreeRow,
};
use crate::workflow_gen::WorkflowFormat;

pub struct WorkflowsSection;

impl StatusSection for WorkflowsSection {
    fn section_id(&self) -> SectionId {
        SectionId::Workflows
    }

    fn label(&self) -> &'static str {
        "Workflows"
    }

    fn prerequisites(&self) -> &[SectionId] {
        &[SectionId::Connections]
    }

    fn health(&self, _snapshot: &StatusSnapshot) -> SectionHealth {
        SectionHealth::Gray
    }

    fn description(&self, _snapshot: &StatusSnapshot) -> String {
        format!("{} export formats", WorkflowFormat::ALL.len())
    }

    fn children(&self, snapshot: &StatusSnapshot) -> Vec<TreeRow> {
        WorkflowFormat::ALL
            .into_iter()
            .map(|fmt| {
                let entry = catalog::entry_for(Vertical::Workflows, fmt.slug());
                let status = entry.as_ref().map(|e| e.status.label()).unwrap_or("Proto");
                let docs_url = entry.and_then(|e| e.docs_url());

                // Primary opens the web Workflows page (where preview/export run);
                // special links to the format's docs when documented.
                let primary = match snapshot.api_port() {
                    Some(port) => StatusAction::OpenWebUiAt {
                        port,
                        route: "/workflows".into(),
                    },
                    None => StatusAction::None,
                };
                let (special, special_meta) = match docs_url {
                    Some(url) => (
                        StatusAction::OpenUrl(url),
                        Some(ActionMeta {
                            title: "Docs",
                            tooltip: "Open this workflow format's documentation",
                        }),
                    ),
                    None => (StatusAction::None, None),
                };

                TreeRow {
                    section_id: SectionId::Workflows,
                    id: format!("workflow-format-{}", fmt.slug()),
                    depth: 1,
                    label: fmt.label().to_string(),
                    description: format!("{status} · .{}", fmt.extension()),
                    icon: StatusIcon::Tool,
                    brand_icon: None,
                    is_header: false,
                    actions: ActionSet {
                        primary,
                        back: StatusAction::None,
                        special,
                        special_meta,
                        refresh: StatusAction::None,
                        refresh_meta: None,
                    },
                    health: SectionHealth::Gray,
                }
            })
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn workflows_section_is_info_only() {
        let snap = StatusSnapshot::from_config(&crate::config::Config::default(), vec![]);
        assert_eq!(WorkflowsSection.health(&snap), SectionHealth::Gray);
        // Gated on Connections: preview/export run against the hosted API.
        assert_eq!(WorkflowsSection.prerequisites(), &[SectionId::Connections]);
    }

    #[test]
    fn workflows_section_lists_every_format() {
        let snap = StatusSnapshot::from_config(&crate::config::Config::default(), vec![]);
        let rows = WorkflowsSection.children(&snap);
        assert_eq!(rows.len(), WorkflowFormat::ALL.len());
        let claude = rows.iter().find(|r| r.label == "Claude Workflow").unwrap();
        assert_eq!(claude.description, "GA · .js");
        // Documented formats expose a Docs link in the special slot.
        assert!(matches!(claude.actions.special, StatusAction::OpenUrl(_)));
    }
}
